### PR TITLE
Igor/linear classifier

### PR DESCRIPTION
**Pull request by Geravant**

### CODE REVIEW

File: Pima indians diabetes and linear classifier/Stochastic Gradient Descent/task.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+### Задание
+
+Реализуйте метод стохастического градиентного спуска для обучения линейного классификатора.
+
+Как и в случае с `GradientDescent` метод `fit` должен
+возвращать значения $Q$ на каждой итерации.
+Чтобы оценить $Q$ на каждой итерации, воспользуйтесь формулой
+$$Q = (1 − \eta)Q + \eta L_i$$
+
+Значение $\eta \in [0, 1]$,
```
Comment (sofiiako) [inline Pima indians diabetes and linear classifier/Stochastic Gradient Descent/task.md:17]: где $L_i$ - это значение функции потерь на данной итерации , а значение $\eta \in [0, 1]$, используемое .....

File: Pima indians diabetes and linear classifier/Stochastic Gradient Descent/task.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+её для определения сходимости не следует. Как мы могли убедиться в предыдущем задании, при неправильном подборе параметров результат достигнут не будет.
+
+Вместо этого предлагается использовать "стратегию оптимиста": сделать ровно `n_iter`
+итераций и надеяться, что за это время стохастический градиентный спуск
+сойдётся.
+
+Параметр `k` определяет размер случайной подвыборки из
+`X`, используемой для вычисления градиента.
+
+В файле `stohastic_gradient-descent.py` пердставлена заготовка реализации. Обратите внимание, что функция для вычисления градиента вынесена отдельно (`calc_grad(self, w, X, y)`), добавлен параметр $\eta$, а главное - в функции `fit` цикл `while` уступил место `for i in range(self.n_iter)`. Это решает проблему неустойчивости `Q`, однако может приводить к худшему приближению результатов. Подобный компромисс времени работы, надежности и устойчивости нередко встречается при настройке алгоритмов машинного обучения.
```
Comment (sofiiako) [inline Pima indians diabetes and linear classifier/Stochastic Gradient Descent/task.md:29]: typo in представлена

File: Pima indians diabetes and linear classifier/Stochastic Gradient Descent/stochastic_gradient_descent.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            q = q * (1 - eta) + loss * eta
+            errors.append(q)
+
+            self.weights -= self.alpha * grad
+
+        return errors
+
+    def calc_grad(self, w, X, y):
+        M = X.dot(w) * y
+        loss, loss_deric = self.loss(M)
```
Comment (sofiiako) [inline Pima indians diabetes and linear classifier/Stochastic Gradient Descent/stochastic_gradient_descent.py:37]: possibly a typo in loss_deric

Comment (sofiiako) [reply Pima indians diabetes and linear classifier/Stochastic Gradient Descent/stochastic_gradient_descent.py:37]: same for all other stochastic_gradient_descent.py files in the project

File: Pima indians diabetes and linear classifier/Read data/stochastic_gradient_descent.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        n = X.shape[1]
+        self.weights = np.random.uniform(-1 / (2 * n), 1 / (2 * n), size=n)
+        errors = []
+
+        q, _ = self.calc_grad(self.weights, X, y)
+        # eta value should be in [0,1] range, this one is chosen arbitrary
+        eta = 1 / len(y)
+        # the algorithm will iterate exactly n_iter times
+        for i in range(self.n_iter):
+            # e need to generate a batch of random k indices from the sample
```
Comment (sofiiako) [inline Pima indians diabetes and linear classifier/Read data/stochastic_gradient_descent.py:26]: we

File: Pima indians diabetes and linear classifier/Read data/stochastic_gradient_descent.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
+import numpy as np
+from loss_functions import sigmoid_loss
+
```
Comment (sofiiako) [inline Pima indians diabetes and linear classifier/Read data/stochastic_gradient_descent.py:3]: PEP 8: E302 expected 2 blank lines, found 1

File: Pima indians diabetes and linear classifier/Stochastic Gradient Descent/task.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                                    ["red", "blue", "green", "magenta", "yellow", "cyan"]):
+                gd = StochasticGradientDescent(alpha=alpha, k=k, n_iter=n_iter)
+```
+Здесь добавлено количество итераций `n_iter = 5000` и еще один внутренний цикл, чтобы посмотреть результаты работы с `k = 1`, `k = 10`, `k = 50`. Для сохранения изображений с каждым рассматриваемым размером подвыборки нужно также модифицировать строки с заданием имен изображений:
+```python
+plt.title("SGD({}, k={})".format(loss.__name__, k))
+plt.legend()
+plt.savefig("sdg-{}-{}.png".format(loss.__name__, k))
+```
+
```
Comment (sofiiako) [inline Pima indians diabetes and linear classifier/Stochastic Gradient Descent/task.md:50]: Apart from that, there were other modifications needed for this to work (maybe we don't need to mention this since these students are probably more advanced, but just in case):
**in utils.py:**
`from stochastic_gradient_descent import StochasticGradientDescent`
**in task.py:**
`from utils import plot_classification`
in main in task.py:
`plot_classification(X, y)`

We can also add something like " Затем запустите task.py. Визуализации работы алгоритма для различных k с использованием различных loss функций появятся слева в Course View."

File: Pima indians diabetes and linear classifier/Gradient Descent/task.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+Полнота показывает количество элементов класса, которые алгоритм сумел верно к нему отнести. В нашем случае это отношение количества анамнезов, которые алгоритм верно отнес к имеющим диабет, к реальному количеству больных диабетом в выборке. Чем больше полнота алгоритма, тем реже он упускает элементы, которые должны были попасть в тот или иной класс.
+</div>
+
+В файле `utils.py` находится функция `train_test_split`, разбивающая заданную выборку на обучающую и контрольную. Также там находится функция `plot_classification`, которая строит график зависимости величины ошибки классификатора от итерации. На нем можно наглядно наблюдать, с какой скоростью изменяется ошибка в зависимости от темпа обучения и функции потерь. Графики для обеих функций будут сохранены как изображения. Как уже упоминалось ранее, градиентный спуск при маленьких значениях темпа обучения сходится довольно медленно, так что будьте готовы к ожиданию результатов работы.
```
Comment (sofiiako) [inline Pima indians diabetes and linear classifier/Gradient Descent/task.md:49]: Some changes to the code needed for this to work. Maybe after "график зависимости величины ошибки классификатора от итерации." we could add:
"Если импортировать эту функцию в task.py, вызвать ее в main и запустить  скрипт, то ....."

File: Pima indians diabetes and linear classifier/Заключение/task.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+В данном уроке мы рассмотрели пример реализации линейного классификатора с использованием стохастического градиентного спуска.
+
+Это наиболее простой в реализации из всех методов локальной оптимизации, однако он является мощным инструментом для решения подобного класса задач. Мы рассмотрели пример классификации для двух классов - болен ли человек диабетом 2 типа или нет. Однако, метод может применяться и для большего количества классов. В таком случае, алгоритм потребует изрядной модификации.
```
Comment (sofiiako) [inline Pima indians diabetes and linear classifier/Заключение/task.md:3]: Этот метод - наиболее простой в реализации из всех методов локальной оптимизации, однако...
 (otherwise it feels like the subject is missing from the sentence)

File: Pima indians diabetes and linear classifier/Заключение/task.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
+В данном уроке мы рассмотрели пример реализации линейного классификатора с использованием стохастического градиентного спуска.
+
+Это наиболее простой в реализации из всех методов локальной оптимизации, однако он является мощным инструментом для решения подобного класса задач. Мы рассмотрели пример классификации для двух классов - болен ли человек диабетом 2 типа или нет. Однако, метод может применяться и для большего количества классов. В таком случае, алгоритм потребует изрядной модификации. 
+
+Для решения задач нет необходимости воспроизводить алгоритм вручную - он реализован, к примеру, в библиотеке [scikit](https://scikit-learn.org/stable/modules/sgd.html). При использовании данной версии необходимо будет передавать функции параметры, разобранные в данном уроке, такие как количество итераций для "стратегии оптимиста" (`n_iter`), или же вид функции потерь (`loss`). Также алгоритм, реализованный в *scikit*, имеет настройки для борьбы с взрывающимися градиентами из коробки.
```
Comment (sofiiako) [inline Pima indians diabetes and linear classifier/Заключение/task.md:5]: для решения реальных задач

**Review by sofiiako — CHANGES_REQUESTED** [review]

File: Pima indians diabetes and linear classifier/Stochastic Gradient Descent/task.md [hunk 10]
```diff
@@ -26,7 +25,7 @@ $$Q = (1 − \eta)Q + \eta L_i$$
 Параметр `k` определяет размер случайной подвыборки из
 `X`, используемой для вычисления градиента.
 
-В файле `stohastic_gradient-descent.py` пердставлена заготовка реализации. Обратите внимание, что функция для вычисления градиента вынесена отдельно (`calc_grad(self, w, X, y)`), добавлен параметр $\eta$, а главное - в функции `fit` цикл `while` уступил место `for i in range(self.n_iter)`. Это решает проблему неустойчивости `Q`, однако может приводить к худшему приближению результатов. Подобный компромисс времени работы, надежности и устойчивости нередко встречается при настройке алгоритмов машинного обучения.
+В файле `stohastic_gradient-descent.py` переставлена заготовка реализации. Обратите внимание, что функция для вычисления градиента вынесена отдельно (`calc_grad(self, w, X, y)`), добавлен параметр $\eta$, а главное - в функции `fit` цикл `while` уступил место `for i in range(self.n_iter)`. Это решает проблему неустойчивости `Q`, однако может приводить к худшему приближению результатов. Подобный компромисс времени работы, надежности и устойчивости нередко встречается при настройке алгоритмов машинного обучения.
```
Comment (sofiiako) [inline Pima indians diabetes and linear classifier/Stochastic Gradient Descent/task.md:28]: переставлена? может представлена?

**Review by sofiiako — APPROVED** [review]
